app/controllers/businessmatching/updateservice/UpdateServiceHelper.scala

```diff
@@ -43,51 +44,52 @@ class UpdateServiceHelper @Inject()(val authConnector: AuthConnector,
+    OptionT(dataCacheConnector.update[BusinessActivities](BusinessActivities.key) {
+      case Some(dcModel) if model.activity.get.equals(AccountancyServices) =>
+        dcModel.accountantForAMLSRegulations(None)
           .whoIsYourAccountant(None)
           .taxMatters(None)
           .copy(hasAccepted = true)
 
       case Some(model) => model
-    }
+    })
   }
 
   def updateSupervision(implicit ac: AuthContext, hc: HeaderCarrier) = {
-    OptionT(dataCacheConnector.fetch[BusinessMatching](BusinessMatching.key)) flatMap { businessMatching =>
-      OptionT.fromOption[Future](businessMatching.activities) flatMap { activities =>
-        (OptionT(dataCacheConnector.fetch[Supervision](Supervision.key)) orElse OptionT.some(Supervision())) flatMap { supervision =>
-          if (activities.businessActivities.intersect(Set(AccountancyServices, TrustAndCompanyServices)).isEmpty) {
-            OptionT.liftF(dataCacheConnector.save[Supervision](Supervision.key, Supervision())) map { _ => Supervision() }
-          } else {
-            OptionT.some(supervision)
-          }
-        }
+    val emptyModel = Supervision(hasAccepted = true)
+
+    for {
+      businessMatching <- OptionT(dataCacheConnector.fetch[BusinessMatching](BusinessMatching.key))
+      activities <- OptionT.fromOption[Future](businessMatching.activities)
+      supervision <- OptionT(dataCacheConnector.fetch[Supervision](Supervision.key)) orElse OptionT.some(emptyModel)
+      newSupervision <- if (activities.businessActivities.intersect(Set(AccountancyServices, TrustAndCompanyServices)).isEmpty) {
+        OptionT.liftF(dataCacheConnector.save[Supervision](Supervision.key, emptyModel)) map { _ => emptyModel }
+      } else {
+        OptionT.some[Future, Supervision](supervision)
       }
-    }
+    } yield newSupervision
   }
 
   def updateHasAcceptedFlag(model: AddServiceFlowModel)(implicit ac: AuthContext, hc: HeaderCarrier) =
     OptionT.liftF(dataCacheConnector.save[AddServiceFlowModel](AddServiceFlowModel.key, model.copy(hasAccepted = true)))
 
-  def updateServicesRegister(activity: BusinessActivity)(implicit ac: AuthContext, hc: HeaderCarrier): Future[Option[ServiceChangeRegister]] =
-    dataCacheConnector.update[ServiceChangeRegister](ServiceChangeRegister.key) {
-      case Some(model@ServiceChangeRegister(Some(activities))) =>
-        model.copy(addedActivities = Some(activities + activity))
-      case _ => ServiceChangeRegister(Some(Set(activity)))
-    }
+  def updateServicesRegister(model: AddServiceFlowModel)(implicit ac: AuthContext, hc: HeaderCarrier):  OptionT[Future, ServiceChangeRegister] =
+    OptionT(dataCacheConnector.update[ServiceChangeRegister](ServiceChangeRegister.key) {
+      case Some(dcModel@ServiceChangeRegister(Some(activities))) =>
+        dcModel.copy(addedActivities = Some(activities +  model.activity.get))
+      case _ => ServiceChangeRegister(Some(Set(model.activity.get)))
+    })
```

Comment:
Could this be rewritten to remove the calls to `.get`? Perhaps as a for-comprehension?

@zatevad